File: capi/supervisor.go
package main

import (
	"context"
	"log"
	"strings"
	"time"

	"github.com/LukasParke/capi/cec"
)

// runCECSupervisor owns the cec.Connection lifecycle: open libcec, find +
// open the adapter, hand the live connection to the global Adapter, drain
// events, and reconnect on any signal from SignalReconnect or after a backoff
// when something fails.
//
// The previous supervisor used many GOTO labels and an unbounded reconnect
// channel drain loop; this version is a flat for/select with explicit
// state transitions and ctx cancellation so the binary can shut down
// cleanly.
func runCECSupervisor(ctx context.Context, deviceName, requestedAdapter string, monitorFromFlag bool) {
	const minBackoff = 3 * time.Second
	const maxBackoff = 60 * time.Second
	backoff := minBackoff

	for {
		// Drain any pending reconnect signals before attempting to open.
		drain(adapter.reconnectSignal())

		conn, opened, err := openCECSession(ctx, deviceName, requestedAdapter, monitorFromFlag)
		if err != nil {
			if ctx.Err() != nil {
				return
			}
			appLog("supervisor", "session start failed: %v (retry in %s)", err, backoff)
			if !sleepCtx(ctx, backoff) {
				return
			}
			backoff = nextBackoff(backoff, maxBackoff)
			continue
		}
		if !opened {
			// ctx was cancelled mid-open
			if conn != nil {
				_ = conn.Close()
			}
			return
		}

		backoff = minBackoff

		adapter.Set(conn)
		globalBusState.setCECReady(true)
		appLog("supervisor", "adapter session ready (monitor=%v)", monitorFromFlag)
		if eventHub != nil {
			eventHub.Publish(CECEvent{
				Type: "adapter_state",
				Data: map[string]interface{}{"state": "connected"},
			})
		}
		signalStewardFull()
		startMQTTFromConfigIfNeeded()

		// Wait for either a reconnect signal or shutdown.
		select {
		case <-ctx.Done():
			adapter.Set(nil)
			_ = conn.Close()
			globalBusState.setCECReady(false)
			setStewardMonitoringState(false)
			globalBusState.setFrameRingCapacity(0)
			return
		case <-adapter.reconnectSignal():
			appLog("supervisor", "reconnect requested; tearing down adapter session")
		}

		if eventHub != nil {
			eventHub.Publish(CECEvent{
				Type: "adapter_state",
				Data: map[string]interface{}{"state": "disconnected"},
			})
		}

		adapter.Set(nil)
		_ = conn.Close()
		globalBusState.setCECReady(false)
		setStewardMonitoringState(false)
		globalBusState.setFrameRingCapacity(0)

		// Brief delay before retrying to avoid hammering libcec/USB.
		if !sleepCtx(ctx, time.Second) {
			return
		}
	}
}

// openCECSession opens a fresh libcec session and tries to attach to the
// adapter. On success, the runCECEventConsumer is started for the new
// connection so events flow into the rest of the service. Returns
// (conn, true, nil) on success, (conn, false, nil) only if ctx was
// cancelled mid-open. On error, returns nil for the connection.
func openCECSession(ctx context.Context, deviceName, requestedAdapter string, monitorFromFlag bool) (*cec.Connection, bool, error) {
	if ctx.Err() != nil {
		return nil, false, nil
	}

	cecCfg := cec.NewConfiguration(deviceName, cec.DeviceTypeRecordingDevice)
	configMu.RLock()
	persisted := currentConfig.CEC
	configMu.RUnlock()
	cecCfg.MonitorOnly = persisted.MonitorOnly
	cecCfg.ActivateSource = persisted.ActivateSource
	cecCfg.WakeDevices = intsToLogicalAddrs(persisted.WakeOnConnect)
	cecCfg.PowerOffDevices = intsToLogicalAddrs(persisted.PowerOffOnDisconnect)

	conn, err := cec.OpenWith(cecCfg, cec.Options{EventBuffer: 512})
	if err != nil {
		return nil, false, err
	}

	go runCECEventConsumer(conn)

	adapterPath, err := pickAdapter(conn, requestedAdapter)
	if err != nil {
		_ = conn.Close()
		return nil, false, err
	}
	appLog("supervisor", "opening CEC adapter: %s", adapterPath)
	if err := conn.OpenAdapter(adapterPath); err != nil {
		_ = conn.Close()
		return nil, false, err
	}

	configMu.RLock()
	busCfg := currentConfig.Bus
	configMu.RUnlock()

	monitor := monitorFromFlag
	if busCfg.MonitorFromConfig != nil {
		monitor = *busCfg.MonitorFromConfig
	}
	if monitor {
		if err := conn.SwitchMonitoring(true); err != nil {
			appLog("supervisor", "SwitchMonitoring enable failed: %v", err)
		} else {
			appLog("cec", "libCEC monitoring enabled: all bus frames forwarded to OnCommand")
		}
	} else {
		_ = conn.SwitchMonitoring(false)
	}
	setStewardMonitoringState(monitor)
	globalBusState.setFrameRingCapacity(busCfg.frameRingSize())

	// Brief settle so the adapter has logical addresses before the first
	// scan. This used to be a flat 2s sleep on the supervisor goroutine;
	// keep that timing but make it cancel-aware.
	if !sleepCtx(ctx, 2*time.Second) {
		_ = conn.Close()
		return nil, false, nil
	}

	return conn, true, nil
}

func pickAdapter(conn *cec.Connection, requested string) (string, error) {
	if requested != "" {
		return requested, nil
	}
	appLog("supervisor", "searching for CEC adapters")
	adapters, err := conn.FindAdapters()
	if err != nil {
		return "", err
	}
	if len(adapters) == 0 {
		return "", ErrAdapterUnavailable
	}
	first := adapters[0]
	if first.Comm != "" && strings.HasPrefix(first.Comm, "/dev/") {
		return first.Comm, nil
	}
	return first.Path, nil
}

// startMQTTFromConfigIfNeeded reads the persisted MQTT config and (re)starts
// the MQTT bridge if a broker is configured and the existing client is not
// connected. Used by the supervisor whenever a fresh adapter session comes up.
func startMQTTFromConfigIfNeeded() {
	configMu.RLock()
	broker := currentConfig.MQTT.Broker
	user := currentConfig.MQTT.User
	pass := currentConfig.MQTT.Pass
	prefix := currentConfig.MQTT.Prefix
	configMu.RUnlock()

	if broker == "" {
		return
	}
	mqttMu.Lock()
	connected := mqttClient != nil && mqttClient.IsConnected()
	mqttMu.Unlock()
	if connected {
		return
	}
	startMQTT(broker, user, pass, prefix)
}

// sleepCtx blocks for d or until ctx is cancelled. Returns false if ctx was
// cancelled (caller should bail out).
func sleepCtx(ctx context.Context, d time.Duration) bool {
	if d <= 0 {
		return ctx.Err() == nil
	}
	t := time.NewTimer(d)
	defer t.Stop()
	select {
	case <-ctx.Done():
		return false
	case <-t.C:
		return true
	}
}

// nextBackoff doubles a backoff value capped at max.
func nextBackoff(cur, max time.Duration) time.Duration {
	next := cur * 2
	if next > max {
		next = max
	}
	return next
}

// drain consumes all currently-buffered values from a non-blocking signal
// channel without blocking.
func drain(ch <-chan struct{}) {
	for {
		select {
		case <-ch:
		default:
			return
		}
	}
}

// intsToLogicalAddrs converts a JSON-friendly []int to []cec.LogicalAddress,
// dropping anything outside 0..14.
func intsToLogicalAddrs(in []int) []cec.LogicalAddress {
	if len(in) == 0 {
		return nil
	}
	out := make([]cec.LogicalAddress, 0, len(in))
	for _, v := range in {
		if v < 0 || v > 14 {
			continue
		}
		out = append(out, cec.LogicalAddress(v))
	}
	return out
}

var _ = log.Println // retain log import for potential future use

File: src/supervisor.rs
//! CEC session supervisor: owns the connection lifecycle — open libcec, find
//! and open the adapter, publish to the shared handle, drain events through
//! the app callback, reconnect on signal or backoff.

use crate::adapter::{AdapterHandle, WaitReason};
use crate::busstate::BusState;
use crate::events::EventHub;
use crate::settings::Settings;
use crate::steward;
use crate::types::{AppEvent, Config};
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::Arc;
use std::time::Duration;
use tracing::{error, info, warn};

pub static SHUTDOWN_FLAG: AtomicBool = AtomicBool::new(false);

const MIN_BACKOFF: Duration = Duration::from_secs(3);
const MAX_BACKOFF: Duration = Duration::from_secs(60);

pub type CecEventSink =
    Arc<dyn Fn(&Arc<crate::cec::Connection>, crate::cec::CecEvent) + Send + Sync>;

pub struct SupervisorDeps {
    pub settings: Arc<Settings>,
    pub adapter: AdapterHandle,
    pub bus: Arc<BusState>,
    pub hub: Arc<EventHub>,
}

/// Runs on a dedicated thread (all libcec calls are blocking). `on_event`
/// receives every CEC event from the live session.
pub fn run_supervisor(
    deps: SupervisorDeps,
    device_name: String,
    requested_adapter: String,
    monitor_from_flag: bool,
    on_event: CecEventSink,
) {
    let mut backoff = MIN_BACKOFF;
    loop {
        if SHUTDOWN_FLAG.load(Ordering::SeqCst) {
            return;
        }
        let cfg: Config = deps.settings.get();
        let cec_cfg = crate::cec::Configuration {
            device_name: device_name.clone(),
            device_type: crate::cec::DeviceType::RECORDING,
            physical_address: 0xFFFF,
            base_device: crate::cec::LogicalAddress::TV,
            hdmi_port: 1,
            monitor_only: cfg.cec.monitor_only,
            activate_source: cfg.cec.activate_source,
            wake_devices: clamp_addrs(&cfg.cec.wake_on_connect),
            power_off_devices: clamp_addrs(&cfg.cec.power_off_on_disconnect),
        };

        let conn = match crate::cec::Connection::open(&cec_cfg) {
            Ok(c) => Arc::new(c),
            Err(e) => {
                error!("session start failed: {e:#} (retry in {backoff:?})");
                if !sleep_std(backoff) {
                    return;
                }
                backoff = next_backoff(backoff);
                continue;
            }
        };

        let picked = if !requested_adapter.is_empty() {
            Ok(requested_adapter.clone())
        } else {
            match conn.find_adapters() {
                Ok(list) if !list.is_empty() => {
                    let first = &list[0];
                    Ok(if first.comm.starts_with("/dev/") {
                        first.comm.clone()
                    } else {
                        first.path.clone()
                    })
                }
                Ok(_) => Err("no CEC adapter found".to_string()),
                Err(e) => Err(format!("find_adapters: {e:#}")),
            }
        };
        let path = match picked {
            Ok(p) => p,
            Err(e) => {
                warn!("{e}");
                let _ = conn.close();
                if !sleep_std(backoff) {
                    return;
                }
                backoff = next_backoff(backoff);
                continue;
            }
        };

        info!("opening CEC adapter: {path}");
        if let Err(e) = conn.open_adapter(&path) {
            error!("open_adapter({path}): {e:#}");
            let _ = conn.close();
            if !sleep_std(backoff) {
                return;
            }
            backoff = next_backoff(backoff);
            continue;
        }

        backoff = MIN_BACKOFF;

        let mut monitor = monitor_from_flag;
        if let Some(v) = cfg.bus.monitor {
            monitor = v;
        }
        if let Err(e) = conn.switch_monitoring(monitor) {
            warn!("switch_monitoring({monitor}): {e:#}");
        }
        steward::set_monitoring(monitor);
        deps.bus.set_frame_ring_capacity(cfg.bus.frame_ring_size());

        deps.adapter.set(Some(conn.clone()));
        deps.bus.set_cec_ready(true);
        info!("adapter session ready (monitor={monitor})");
        publish_state(&deps.hub, "connected");

        // Event consumer for this session.
        let mut rx = conn.subscribe_events();
        let consumer_conn = conn.clone();
        let on_event = on_event.clone();
        std::thread::Builder::new()
            .name("cec-events".into())
            .spawn(move || loop {
                match rx.blocking_recv() {
                    Ok(ev) => on_event(&consumer_conn, ev),
                    Err(tokio::sync::broadcast::error::RecvError::Lagged(n)) => {
                        warn!("cec event consumer lagged, dropped {n}");
                    }
                    Err(tokio::sync::broadcast::error::RecvError::Closed) => return,
                }
            })
            .expect("spawn cec-events");

        match deps
            .adapter
            .wait_for(&SHUTDOWN_FLAG, Duration::from_secs(u64::MAX / 2))
        {
            WaitReason::Timeout => continue,
            reason => {
                publish_state(&deps.hub, "disconnected");
                deps.adapter.set(None);
                deps.bus.set_cec_ready(false);
                steward::set_monitoring(false);
                deps.bus.set_frame_ring_capacity(0);
                let _ = conn.close();
                match reason {
                    WaitReason::Shutdown => return,
                    WaitReason::Reconnect => {
                        info!("reconnect requested; tearing down adapter session");
                        if !sleep_std(Duration::from_secs(1)) {
                            return;
                        }
                    }
                    WaitReason::Timeout => unreachable!(),
                }
            }
        }
    }
}

fn publish_state(hub: &Arc<EventHub>, state: &str) {
    hub.publish(AppEvent::new(
        crate::types::event_type::ADAPTER_STATE,
        serde_json::json!({ "state": state }),
    ));
}

/// Shutdown-aware sleep for supervisor threads.
pub fn sleep_std(d: Duration) -> bool {
    let deadline = std::time::Instant::now() + d;
    while std::time::Instant::now() < deadline {
        if SHUTDOWN_FLAG.load(Ordering::SeqCst) {
            return false;
        }
        std::thread::sleep(
            Duration::from_millis(100)
                .min(deadline.saturating_duration_since(std::time::Instant::now())),
        );
    }
    !SHUTDOWN_FLAG.load(Ordering::SeqCst)
}

fn next_backoff(cur: Duration) -> Duration {
    (cur * 2).min(MAX_BACKOFF)
}

fn clamp_addrs(v: &[i32]) -> Vec<u8> {
    v.iter()
        .filter(|x| **x >= 0 && **x <= 14)
        .map(|x| *x as u8)
        .collect()
}
